File: tests/test_sam_protocol.py
import unittest

from i2pchat.sam.errors import CantReachPeer, DuplicatedId, InvalidId, LegacySAMException, ProtocolError
from i2pchat.sam.protocol import expect_ok, parse_reply_line, redact_sam_line


class SamProtocolTests(unittest.TestCase):
    def test_parse_reply_line_extracts_command_topic_and_fields(self) -> None:
        reply = parse_reply_line(b"HELLO REPLY RESULT=OK VERSION=3.1\n")
        self.assertEqual(reply.command, "HELLO")
        self.assertEqual(reply.topic, "REPLY")
        self.assertEqual(reply.fields["RESULT"], "OK")
        self.assertEqual(reply.fields["VERSION"], "3.1")

    def test_parse_reply_line_rejects_empty_line(self) -> None:
        with self.assertRaises(ProtocolError):
            parse_reply_line(b"\n")

    def test_parse_reply_line_rejects_single_token(self) -> None:
        with self.assertRaises(ProtocolError) as ctx:
            parse_reply_line(b"HELLO\n")
        self.assertIn("Malformed", ctx.exception.message)

    def test_parse_reply_line_rejects_whitespace_only(self) -> None:
        with self.assertRaises(ProtocolError) as ctx:
            parse_reply_line(b"   \n")
        self.assertIn("Empty", ctx.exception.message)

    def test_parse_reply_line_accepts_two_token_minimal_reply(self) -> None:
        reply = parse_reply_line(b"STREAM STATUS\n")
        self.assertEqual(reply.command, "STREAM")
        self.assertEqual(reply.topic, "STATUS")
        self.assertEqual(reply.fields, {})

    def test_expect_ok_returns_reply_on_success(self) -> None:
        reply = parse_reply_line(b"STREAM STATUS RESULT=OK\n")
        self.assertIs(expect_ok(reply), reply)

    def test_expect_ok_accepts_dest_reply_pub_priv_without_result(self) -> None:
        """i2pd may omit RESULT=OK on DEST GENERATE success (DEST REPLY PUB=… PRIV=…)."""
        line = (
            b"DEST REPLY PUB=aaa PRIV=bbb\n"
        )
        reply = parse_reply_line(line)
        self.assertIs(expect_ok(reply), reply)

    def test_expect_ok_accepts_session_status_destination_without_result(self) -> None:
        """i2pd may omit RESULT=OK on successful SESSION CREATE (STATUS + DESTINATION=…)."""
        reply = parse_reply_line(b"SESSION STATUS DESTINATION=TRANSIENT\n")
        self.assertIs(expect_ok(reply), reply)

    def test_expect_ok_maps_known_sam_errors(self) -> None:
        cant_reach = parse_reply_line(
            b"STREAM STATUS RESULT=CANT_REACH_PEER MESSAGE=offline\n"
        )
        with self.assertRaises(CantReachPeer):
            expect_ok(cant_reach)

        invalid_id = parse_reply_line(
            b"STREAM STATUS RESULT=INVALID_ID MESSAGE=unknown-session\n"
        )
        with self.assertRaises(InvalidId):
            expect_ok(invalid_id)

    def test_expect_ok_rejects_missing_result(self) -> None:
        reply = parse_reply_line(b"STREAM STATUS MESSAGE=no-result\n")
        with self.assertRaises(ProtocolError):
            expect_ok(reply)

    def test_expect_ok_maps_duplicated_id(self) -> None:
        reply = parse_reply_line(
            b"SESSION STATUS RESULT=DUPLICATED_ID MESSAGE=dup\n"
        )
        with self.assertRaises(DuplicatedId):
            expect_ok(reply)

    def test_expect_ok_maps_unknown_result_to_legacy(self) -> None:
        reply = parse_reply_line(b"STREAM STATUS RESULT=NO_SUCH_THING MESSAGE=x\n")
        with self.assertRaises(LegacySAMException) as ctx:
            expect_ok(reply)
        self.assertEqual(ctx.exception.result, "NO_SUCH_THING")

    def test_redact_sam_line_masks_private_key_material(self) -> None:
        raw = "DEST REPLY PUB=pubvalue PRIV=secretkey RESULT=OK"
        redacted = redact_sam_line(raw)
        self.assertNotIn("secretkey", redacted)
        self.assertIn("PUB=pubvalue", redacted)  # PUB is public material and stays
        self.assertIn("RESULT=OK", redacted)

    def test_parse_reply_line_does_not_retain_priv_in_raw_line(self) -> None:
        reply = parse_reply_line(b"DEST REPLY PUB=aaa PRIV=supersecret\n")
        self.assertNotIn("supersecret", reply.raw_line or "")
        # Programmatic extraction still works via fields.
        self.assertEqual(reply.fields["PRIV"], "supersecret")

    def test_expect_ok_error_does_not_leak_destination_in_raw_line(self) -> None:
        reply = parse_reply_line(
            b"SESSION STATUS RESULT=I2P_ERROR DESTINATION=longsecretdest MESSAGE=boom\n"
        )
        with self.assertRaises(LegacySAMException) as ctx:
            expect_ok(reply)
        self.assertNotIn("longsecretdest", ctx.exception.raw_line or "")
        self.assertNotIn("longsecretdest", repr(ctx.exception))


if __name__ == "__main__":
    unittest.main()

File: cpp/tests/test_sam_protocol.cpp
#include <catch2/catch_test_macros.hpp>

#include "i2pchat/sam/protocol.hpp"
#include "vectors.hpp"

using namespace i2pchat;
using i2pchat::testing::load_vector;

TEST_CASE("SAM command builders reproduce the reference bytes", "[sam][vectors]") {
    const auto document = load_vector("sam");

    for (const auto& entry : document.at("builders")) {
        const auto name = entry.at("name").get<std::string>();
        const auto expected = entry.at("payload").get<std::string>();
        if (name == "hello") {
            CHECK(sam::build_hello() == expected);
        } else if (name == "dest_generate") {
            CHECK(sam::build_dest_generate() == expected);
        }
    }
}

TEST_CASE("SESSION CREATE keeps the unconditional space before options",
          "[sam]") {
    // The reference builder always emits a space after DESTINATION, so with no
    // options the line ends with a trailing space. Routers accept it; matching
    // it keeps captured traffic identical between implementations.
    CHECK(sam::build_session_create("STREAM", "sess-1", "TRANSIENT") ==
          "SESSION CREATE STYLE=STREAM ID=sess-1 DESTINATION=TRANSIENT \n");

    CHECK(sam::build_session_create("STREAM", "sess-1", "TRANSIENT", {},
                                    sam::kSigTypeEd25519) ==
          "SESSION CREATE STYLE=STREAM ID=sess-1 DESTINATION=TRANSIENT "
          "SIGNATURE_TYPE=7\n");

    CHECK(sam::build_session_create("STREAM", "sess-1", "TRANSIENT",
                                    {{"inbound.length", "3"}}, sam::kSigTypeEd25519) ==
          "SESSION CREATE STYLE=STREAM ID=sess-1 DESTINATION=TRANSIENT "
          "SIGNATURE_TYPE=7 inbound.length=3\n");
}

TEST_CASE("stream commands are built as the router expects", "[sam]") {
    CHECK(sam::build_stream_connect("sess-1", "dest-b64") ==
          "STREAM CONNECT ID=sess-1 DESTINATION=dest-b64 SILENT=false\n");
    CHECK(sam::build_stream_accept("sess-1") ==
          "STREAM ACCEPT ID=sess-1 SILENT=false\n");
    CHECK(sam::build_naming_lookup("example.i2p") ==
          "NAMING LOOKUP NAME=example.i2p\n");
    // STREAM FORWARD carries a trailing space in the reference builder too.
    CHECK(sam::build_stream_forward("sess-1", 1234) ==
          "STREAM FORWARD ID=sess-1 PORT=1234 \n");
}

TEST_CASE("builders reject injected control characters", "[sam]") {
    // Without this a caller-supplied name could smuggle an extra SAM command.
    CHECK_THROWS_AS(sam::build_naming_lookup("host\nSESSION CREATE"), sam::SamError);
    CHECK_THROWS_AS(sam::build_stream_connect("sess\r\n1", "dest"), sam::SamError);
    CHECK_THROWS_AS(sam::build_session_create("STREAM", "id with space", "TRANSIENT"),
                    sam::SamError);
    CHECK_THROWS_AS(sam::build_session_create("GOSSIP", "id", "TRANSIENT"),
                    sam::SamError);
    CHECK_THROWS_AS(sam::build_hello("three", "3.2"), sam::SamError);
    CHECK_THROWS_AS(sam::build_stream_forward("sess-1", 0), sam::SamError);
    CHECK_THROWS_AS(sam::build_stream_forward("sess-1", 70000), sam::SamError);
}

TEST_CASE("reply parsing upper-cases the command, topic and keys", "[sam]") {
    const sam::SamReply reply =
        sam::parse_reply_line("session status result=OK destination=abc\n");
    CHECK(reply.command == "SESSION");
    CHECK(reply.topic == "STATUS");
    CHECK(reply.field("RESULT") == "OK");
    // Values keep their case; only keys are normalized.
    CHECK(reply.field("destination") == "abc");
}

TEST_CASE("reply parsing rejects empty and malformed lines", "[sam]") {
    CHECK_THROWS_AS(sam::parse_reply_line(""), sam::SamError);
    CHECK_THROWS_AS(sam::parse_reply_line("   \n"), sam::SamError);
    CHECK_THROWS_AS(sam::parse_reply_line("HELLO"), sam::SamError);
}

TEST_CASE("private key material is redacted from reply lines", "[sam]") {
    // A DEST REPLY carries the profile's private identity key; it must never
    // survive into a log line or an exception message.
    const sam::SamReply reply =
        sam::parse_reply_line("DEST REPLY PUB=publicvalue PRIV=secretkeymaterial");
    CHECK(reply.raw_line.find("secretkeymaterial") == std::string::npos);
    CHECK(reply.raw_line.find("<redacted:") != std::string::npos);
    // The parsed field is still available to the caller that needs it.
    CHECK(reply.field("PRIV") == "secretkeymaterial");

    const sam::SamReply session =
        sam::parse_reply_line("SESSION STATUS RESULT=OK DESTINATION=longdestination");
    CHECK(session.raw_line.find("longdestination") == std::string::npos);
}

TEST_CASE("expect_ok accepts i2pd's implicit successes", "[sam]") {
    // i2pd omits RESULT=OK on a successful DEST GENERATE.
    const sam::SamReply dest = sam::parse_reply_line("DEST REPLY PUB=pub PRIV=priv");
    CHECK_NOTHROW(sam::expect_ok(dest));

    // And often on SESSION CREATE.
    const sam::SamReply session = sam::parse_reply_line("SESSION STATUS DESTINATION=d");
    CHECK_NOTHROW(sam::expect_ok(session));

    const sam::SamReply ok = sam::parse_reply_line("HELLO REPLY RESULT=OK VERSION=3.1");
    CHECK_NOTHROW(sam::expect_ok(ok));
}

TEST_CASE("expect_ok maps SAM result codes to error kinds", "[sam]") {
    const auto check_kind = [](const std::string& line, sam::SamErrorKind expected) {
        const sam::SamReply reply = sam::parse_reply_line(line);
        try {
            sam::expect_ok(reply);
            FAIL("expected expect_ok to throw for: " + line);
        } catch (const sam::SamError& error) {
            CHECK(error.kind() == expected);
        }
    };

    check_kind("STREAM STATUS RESULT=CANT_REACH_PEER", sam::SamErrorKind::CantReachPeer);
    check_kind("SESSION STATUS RESULT=DUPLICATED_ID", sam::SamErrorKind::DuplicatedId);
    check_kind("SESSION STATUS RESULT=DUPLICATED_DEST", sam::SamErrorKind::DuplicatedDest);
    check_kind("STREAM STATUS RESULT=I2P_ERROR", sam::SamErrorKind::I2pError);
    check_kind("STREAM STATUS RESULT=INVALID_KEY", sam::SamErrorKind::InvalidKey);
    check_kind("NAMING REPLY RESULT=KEY_NOT_FOUND", sam::SamErrorKind::KeyNotFound);
    check_kind("STREAM STATUS RESULT=PEER_NOT_FOUND", sam::SamErrorKind::PeerNotFound);
    check_kind("STREAM STATUS RESULT=TIMEOUT", sam::SamErrorKind::Timeout);
    check_kind("STREAM STATUS RESULT=SOMETHING_NEW", sam::SamErrorKind::Unknown);

    // A reply with no RESULT and no implicit-success payload is a protocol error.
    check_kind("STREAM STATUS FOO=bar", sam::SamErrorKind::Protocol);
}
